File: 2561-rearranging-fruits/2561-rearranging-fruits.cpp
#include <vector>
#include <map>
#include <algorithm>
#include <climits>

class Solution {
public:
    long long minCost(std::vector<int>& basket1, std::vector<int>& basket2) {
        std::map<int, int> total_counts;
        for (int fruit : basket1) total_counts[fruit]++;
        for (int fruit : basket2) total_counts[fruit]++;

        long long min_val = LLONG_MAX;
        for (auto const& [fruit, count] : total_counts) {
            if (count % 2 != 0) return -1;
            min_val = std::min(min_val, (long long)fruit);
        }
        
        std::vector<long long> fruits_to_swap;
        std::map<int, int> count1;
        for (int fruit : basket1) count1[fruit]++;
        
        for (auto const& [fruit, total_count] : total_counts) {
            int diff = count1[fruit] - (total_count / 2);
            for (int i = 0; i < abs(diff); ++i) {
                fruits_to_swap.push_back(fruit);
            }
        }
        
        std::sort(fruits_to_swap.begin(), fruits_to_swap.end());
        
        long long total_cost = 0;
        int swaps_to_make = fruits_to_swap.size() / 2;
        for (int i = 0; i < swaps_to_make; ++i) {
            total_cost += std::min(fruits_to_swap[i], 2 * min_val);
        }
        
        return total_cost;
    }
};

File: 2561-rearranging-fruits/2561-rearranging-fruits.py
from collections import Counter

class Solution:
    def minCost(self, basket1: list[int], basket2: list[int]) -> int:
        total_counts = Counter(basket1) + Counter(basket2)
        
        for count in total_counts.values():
            if count % 2 != 0:
                return -1
        
        fruits_to_swap = []
        count1 = Counter(basket1)
        for fruit, total_count in total_counts.items():
            target = total_count // 2
            diff = count1.get(fruit, 0) - target
            
            # if diff > 0, b1 has a surplus. if diff < 0, b2 has a surplus.
            # abs(diff) is the number of items of this fruit type in the wrong basket.
            # since each misplaced fruit corresponds to another, we add half.
            for _ in range(abs(diff)):
                fruits_to_swap.append(fruit)

        fruits_to_swap.sort()
        
        min_val = min(total_counts.keys())
        total_cost = 0
        swaps_to_make = len(fruits_to_swap) // 2
        for i in range(swaps_to_make):
            # Cost is min(direct swap cost, helper swap cost)
            total_cost += min(fruits_to_swap[i], 2 * min_val)
            
        return total_cost